src/print_graph.cc

```diff
@@ -24,133 +22,197 @@ namespace detail {
+			const auto reduction = rm.get_reduction(rrcmd->get_rid());
+			const auto req = GridRegion<3>{{1, 1, 1}};
+			fmt::format_to(std::back_inserter(label), "<b>reduction</b> R{}<br/>B{} {}", rrcmd->get_rid(), reduction.output_buffer_id, req);
+		} else if(const auto hcmd = dynamic_cast<const horizon_command*>(&cmd)) {
+			label += "<b>horizon</b>";
 		} else {
 			assert(!"Unkown command");
-			return fmt::format("[{}] UNKNOWN\\n{}", cmd->get_cid(), cmd->debug_label);
+			label += "<b>unknown</b>";
 		}
-		return label;
-	}
 
-	std::string print_command_graph(const command_graph& cdag, const task_manager& tm) {
-		std::ostringstream main_ss;
-		std::unordered_map<task_id, std::ostringstream> task_subgraph_ss;
+		if(const auto tcmd = dynamic_cast<const task_command*>(&cmd)) {
+			const auto& tsk = *tm.get_task(tcmd->get_tid());
 
-		const auto write_vertex = [&](std::ostream& out, abstract_command* cmd) {
-			const char* colors[] = {"black", "crimson", "dodgerblue4", "goldenrod", "maroon4", "springgreen2", "tan1", "chartreuse2"};
+			auto reduction_init_mode = access_mode::discard_write;
+			auto execution_range = subrange<3>{tsk.get_global_offset(), tsk.get_global_size()};
+			if(const auto ecmd = dynamic_cast<const execution_command*>(&cmd)) {
+				if(ecmd->is_reduction_initializer()) { reduction_init_mode = cl::sycl::access::mode::read_write; }
+				execution_range = ecmd->get_execution_range();
+			}
+
+			for(auto rid : tsk.get_reductions()) {
+				auto reduction = rm.get_reduction(rid);
 
-			std::unordered_map<std::string, std::string> props;
-			props["label"] = "\"" + get_command_label(cmd) + "\"";
-			props["fontcolor"] = colors[cmd->get_nid() % (sizeof(colors) / sizeof(char*))];
-			if(isa<task_command>(cmd)) { props["shape"] = "box"; }
+				auto rmode = cl::sycl::access::mode::discard_write;
+				if(reduction.initialize_from_buffer) { rmode = reduction_init_mode; }
 
-			out << cmd->get_cid();
-			out << "[";
-			for(const auto& it : props) {
-				out << " " << it.first << "=" << it.second;
+				const auto bid = reduction.output_buffer_id;
+				const auto req = GridRegion<3>{{1, 1, 1}};
+				fmt::format_to(std::back_inserter(label), "<br/>(R{}) <i>{}</i> B{} {}", rid, detail::access::mode_traits::name(rmode), bid, req);
+			}
+
+			const auto& bam = tsk.get_buffer_access_map();
+			for(const auto bid : bam.get_accessed_buffers()) {
+				for(const auto mode : bam.get_access_modes(bid)) {
+					const auto req = bam.get_requirements_for_access(bid, mode, tsk.get_dimensions(), execution_range, tsk.get_global_size());
+					// While uncommon, we do support chunks that don't require access to a particular buffer at all.
+					if(!req.empty()) { fmt::format_to(std::back_inserter(label), "<br/><i>{}</i> B{} {}", detail::access::mode_traits::name(mode), bid, req); }
+				}
 			}
-			out << "];";
+
+			for(const auto& [hoid, order] : tsk.get_side_effect_map()) {
+				fmt::format_to(std::back_inserter(label), "<br/><i>affect</i> H{}", hoid);
```

Comment:
Not sure if worthwhile, but there seems to be quite a bit of code duplication with `get_task_label`.

Reply:
You're right! I've deduplicated this.